packages/react-hooks/src/ConnectionProvider.tsx

```diff
@@ -15,14 +15,71 @@ import {
   addRecord,
 } from './recordUtils'
 
+export type useConnectionsOptions = {
+  excludedTypes?: [ConnectionType | string]
+  connectionState?: DidExchangeState
+}
+
 const ConnectionContext = createContext<RecordsState<ConnectionRecord> | undefined>(undefined)
 
-export const useConnections = () => {
+/**
+ * This method retreives the connection context for the current agent.
+ * From this you can access all connection records for the agent.
+ * @param options options for useConnections hook, lets us filter out specific types and limit states
+ * @returns a connection context containing information about the agents connections
+ */
+export const useConnections = (options: useConnectionsOptions = {}) => {
   const connectionContext = useContext(ConnectionContext)
   if (!connectionContext) {
     throw new Error('useConnections must be used within a ConnectionContextProvider')
   }
-  return connectionContext
+
+  let connections = connectionContext.records
+
+  // Only run this code if we don't have options.excludedTypes to limit ammount of loops
+  if (options.connectionState && !options.excludedTypes) {
+    const filteredByState = useMemo(
+      () => connections.filter((c: ConnectionRecord) => c.state === options.connectionState),
+      [connections, options.connectionState]
+    )
+    connections = filteredByState
+  }
+
+  if (options.excludedTypes) {
+    const filteredByType = useMemo(
+      () =>
+        connections.filter((record: ConnectionRecord) => {
+          const recordTypes = (record.getTag('connectionType') as [string] | null) || []
+          // Filter out by state if needed
+          if (options.connectionState && options.connectionState !== record.state) {
+            return false
+          }
+          for (const type of recordTypes) {
+            if (options.excludedTypes?.includes(type)) return false
+          }
+          return true
+        }),
+      [connections, options.excludedTypes]
+    )
+    connections = filteredByType
+  }
```

Comment:
If we write it like this, it's only a single filter, which scales a bit better if new filter type are added (just add another if statement):

```suggestion
  // Only run this code if we don't have options.excludedTypes or options.connectionState to limit amount of loops
  if (!options.connectionState && !options.excludedTypes) return connections
  
  // Filter connections based on provided filters
  return useMemo(() => connections.filter((connection) => {
   // By default we include this connection
   let valid = true
   
   
   // Filter by state (if connectionState is defined)
    if (options.connectionState) valid = c.state === options.connectionState
    
    // Exclude records with certain connection types (if defined)
    const connectionTypes = record.getTag('connectionType') as string[] | null
    if (options.excludedTypes && recordTypes) {
      valid = connectionTypes.some(connectionType => options.excludedTypes.includes(connectionType))
    }
    
    return valid
  }), [connections, options.connectionState, options.excludedTypes])
```

Reply:
Also I'm not sure whether this will work, as we're using a conditional hook here, which is not allowed according to the rules of hooks. https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

So we should probably model it like this:

```ts  
  return useMemo(() => {
      // do not filter if not filter options are provided to save on a loop
      if (!options.connectionState && !options.excludedTypes) return connections

       return connections.filter((connection) => {
	   // By default we include this connection
	   let valid = true
	   
	   
	   // Filter by state (if connectionState is defined)
	    if (options.connectionState) valid = c.state === options.connectionState
	    
	    // Exclude records with certain connection types (if defined)
	    const connectionTypes = record.getTag('connectionType') as string[] | null
	    if (options.excludedTypes && recordTypes) {
	      valid = connectionTypes.some(connectionType => options.excludedTypes.includes(connectionType))
	    }
	    
	    return valid
	  })
     }, [connections, options.connectionState, options.excludedTypes])
```

Reply:
Bump on this last comment @ryankoch13 ^